File: 2441-largest-positive-integer-that-exists-with-its-negative/2441-largest-positive-integer-that-exists-with-its-negative.cpp
class Solution {
public:
    int findMaxK(vector<int>& nums) {
        if (nums.size() < 2) {
            /// length == 1 -> not possible
            return -1;
        }
        sort(nums.begin(), nums.end());
        
        if (nums[0] > 0) {
            /// nums[0] after sort is > 0 -> no {-k,k} set possible
            return -1;
        }
        
        int l = 0, r = nums.size() - 1;
        while (l < r) {
            if (nums[l] + nums[r] == 0) {
                return nums[r];
            } else if (nums[l] + nums[r] > 0) {
                r--;
            } else {
                /// sum < 0 
                l++;
            }
        }
        
        return -1;
    }
};

File: 2441-largest-positive-integer-that-exists-with-its-negative/2441-largest-positive-integer-that-exists-with-its-negative.rs
impl Solution {
    pub fn find_max_k(nums: Vec<i32>) -> i32 {
        let mut nnums = nums.clone();
        let n = nums.len();
        
        if n < 2 {
            return -1; 
        }
        
        nnums.sort();
        
        if nnums[0] > 0 {
            return -1;
        }
        
        let (mut left, mut right) = (0, n - 1);
        
        while left < right {
            let sum = nnums[left as usize] + nnums[right as usize];
            if sum == 0 {
                return nnums[right as usize];
            } else if sum > 0 {
                right -= 1;
            } else {
                left += 1;
            }
        }
        
        -1
    }
}